### PR TITLE
BUG: Fix segfault in GroupBy.count and DataFrame.count

### LINKED ISSUE #21824: Segfault on clean-up with count example from docstrings
In https://github.com/pandas-dev/pandas/pull/19952 I have problems with a segfault when running doctests, and I could isolate at least the following code snippet that segfaults on clean-up (so only when exiting the process; an explicit del + gc.collect does not yet trigger the segfault).

```
import numpy as  np
import pandas as pd

df = pd.DataFrame({"Person": ["John", "Myla", None, "John", "Myla"],
                   "Age": [24., np.nan, 21., 33, 26],
                   "Single": [False, True, True, True, False]})
                  
res = df.set_index(["Person", "Single"]).count(level="Person")
```

### LINKED ISSUE #32841: SegmentationFault/BusError: Groupby + count
#### Code Sample

```python
import pandas as pd
import numpy as np
df = pd.DataFrame({"A": [0, 1, np.NaN], "B": [1, 2, 3]})
df.groupby(["A"]).count()
```

#### Problem description

This randomly causes a SegFault or Bus error.

With bounds checking enabled I get the following IndexError.

https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/lib.pyx#L778
```python
@cython.boundscheck(True)
@cython.wraparound(False)
def count_level_2d(ndarray[uint8_t, ndim=2, cast=True] mask,
...
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/pandas/pandas/core/groupby/generic.py", line 1784, in count
    blocks = [make_block(val, placement=loc) for val, loc in zip(counted, locs)]
  File "/pandas/pandas/core/groupby/generic.py", line 1784, in <listcomp>
    blocks = [make_block(val, placement=loc) for val, loc in zip(counted, locs)]
  File "/pandas/pandas/core/groupby/generic.py", line 1782, in <genexpr>
    lib.count_level_2d(x, labels=ids, max_bin=ngroups, axis=1) for x in vals
  File "pandas/_libs/lib.pyx", line 803, in pandas._libs.lib.count_level_2d
    counts[i, labels[j]] += mask[i, j]
IndexError: Out of bounds on buffer access (axis 1)
```

#### Expected Output

```python
     B
A     
0.0  1
1.0  1
```

### Note

This issue is related to https://github.com/pandas-dev/pandas/issues/21824. The Segfault happens in the same function, but the call stack is different.

Code sample from https://github.com/pandas-dev/pandas/issues/21824:
```python
import numpy as  np
import pandas as pd

df = pd.DataFrame({"Person": ["John", "Myla", None, "John", "Myla"],
                   "Age": [24., np.nan, 21., 33, 26],
                   "Single": [False, True, True, True, False]})
                  
res = df.set_index(["Person", "Single"]).count(level="Person")
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/pandas/pandas/core/frame.py", line 7786, in count
    return self._count_level(level, axis=axis, numeric_only=numeric_only)
  File "/pandas/pandas/core/frame.py", line 7842, in _count_level
    counts = lib.count_level_2d(mask, level_codes, len(level_index), axis=0)
  File "pandas/_libs/lib.pyx", line 796, in pandas._libs.lib.count_level_2d
    counts[labels[i], j] += mask[i, j]
IndexError: Out of bounds on buffer access (axis 0)
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 981252e7adb74b3aca2d9dfd075b0b64c3552975
python           : 3.7.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
Version          : Darwin Kernel Version 18.7.0: Thu Jan 23 06:52:12 PST 2020; root:xnu-4903.278.25~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.1.0.dev0+870.g981252e7a
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.0.0.post20200311
Cython           : 0.29.15
pytest           : 5.4.1
hypothesis       : 5.6.0
sphinx           : 2.4.4
blosc            : 1.8.3
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : 1.3.2
fastparquet      : 0.3.3
gcsfs            : None
matplotlib       : 3.2.0
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.1
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
pyxlsb           : None
s3fs             : 0.4.0
scipy            : 1.4.1
sqlalchemy       : 1.3.15
tables           : 3.6.1
tabulate         : 0.8.6
xarray           : 0.15.0
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.48.0

</details>

### PR DESCRIPTION
**Pull request by tv3141**
- [x] closes #32841 
- [x] closes #21824
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -793,14 +793,16 @@ def count_level_2d(ndarray[uint8_t, ndim=2, cast=True] mask,
...
+                        counts[labels[i], j] += 1
 
     else:  # axis == 1
         counts = np.zeros((n, max_bin), dtype='i8')
         with nogil:
             for i in range(n):
                 for j in range(k):
-                    counts[i, labels[j]] += mask[i, j]
+                    if mask[i, j]:
+                        counts[i, labels[j]] += 1
```
Comment (tv3141) [inline pandas/_libs/lib.pyx]: `count()` is handling `-1` by masking them.

```python
    def count(self):
        """
        Compute count of group, excluding missing values.
        Returns
        -------
        DataFrame
            Count of values within each group.
        """
        data = self._get_data_to_aggregate()
        ids, _, ngroups = self.grouper.group_info
        mask = ids != -1

        # TODO(2DEA): reshape would not be necessary with 2D EAs
        vals = ((mask & ~isna(blk.values).reshape(blk.shape)) for blk in data.blocks)
        locs = (blk.mgr_locs for blk in data.blocks)

        counted = (
            lib.count_level_2d(x, labels=ids, max_bin=ngroups, axis=1) for x in vals
        )
```

`count_level_2d` was designed to ignore masked values.
This is essentially reverting this change:
https://github.com/pandas-dev/pandas/commit/d968aab3da13f68fabbb05f8eba12927d8bc5836#diff-8fa7422077eecd07a4006c11c78fc93aL1265-L1268

**Comment by tv3141** [comment]
So far, this PR does not fix the related https://github.com/pandas-dev/pandas/issues/21824.

One option would be to mask rows `-1` as well in the calling code as well. But I think it would be good to prevent SegFaults in `count_level_2d` itself, especially if bound checking should be disabled for performance.

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -775,7 +775,7 @@ def get_level_sorter(const int64_t[:] label, const int64_t[:] starts):
     return out
 
 
-@cython.boundscheck(False)
+@cython.boundscheck(True)
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx]: Not sure if you disabled this temporarily for debugging purposes but want to revert this; this has performance implications re-enabling

**Review by WillAyd — COMMENTED** [review]
Thanks for taking a look; on first pass I think this code is essentially the same thing though?

Do you know which array access / index is actually causing the OOB error?

**Comment by tv3141** [comment]
> Thanks for taking a look; on first pass I think this code is essentially the same thing though?
> 
> Do you know which array access / index is actually causing the OOB error?

`labels` contains `-1`s when the groupby column contains `NaN`s. 

The calling function `count()` creates `mask` for `-1`s, and `count_level_2d` would ignore masked values. This changed with https://github.com/pandas-dev/pandas/commit/d968aab3da13f68fabbb05f8eba12927d8bc5836#diff-8fa7422077eecd07a4006c11c78fc93aL1265-L1268, since then out-of-bounds access on `counts` can happen.

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (tv3141) [reply pandas/_libs/lib.pyx]: I disabled this to get a reproducible error + traceback, rather than random segfaults/bus errors.
On the other hand, it is still very easy to do out-of-bounds access in `count_level_2d`. #21824 still does, this PR does not fix that issue.

Ideally, the negative indexes should be handled in `count_level_2d`, so that disabling bounds checking does not lead to random segfaults.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -7820,13 +7820,21 @@ def _count_level(self, level, axis=0, numeric_only=False):
                 f"Can only count levels on hierarchical {self._get_axis_name(axis)}."
             )
 
+        # Mask NaNs: Mask rows where the index level is NaN and all values in
+        # the DataFrame that are NaN
```
Comment (jreback) [inline pandas/core/frame.py]: you can remove this if/else as they are the same

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine, merge master to fix the docs fails, also comment. ping on green.

**Review by jreback — CHANGES_REQUESTED** [review]
actually need a whatsnew note, bug fixes for groupby.

File: pandas/_libs/lib.pyx [hunk 4]
```diff
@@ -793,14 +793,16 @@ def count_level_2d(ndarray[uint8_t, ndim=2, cast=True] mask,
         with nogil:
             for i in range(n):
                 for j in range(k):
-                    counts[labels[i], j] += mask[i, j]
+                    if mask[i, j]:
+                        counts[labels[i], j] += 1
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx]: Sorry if missed before but can you share why this change is required? Isn't the mask only ever 0 or 1?

Comment (tv3141) [reply pandas/_libs/lib.pyx]: The segfault is caused by out-of-bounds access on `counts`, when `labels[i]` is `-1`. If it is checked first whether a value should be counted then out-of-bounds access does not happen.

This is also how the code was written originally:
https://github.com/pandas-dev/pandas/commit/d968aab3da13f68fabbb05f8eba12927d8bc5836#diff-8fa7422077eecd07a4006c11c78fc93aL1265-L1268

That means that `count_level_2d` requires `mask` to be `0` 
1. on columns/rows that should not be counted, because the counted-over row index or column index value is `NaN`.
1. for data values that are `NaN`

Comment (WillAyd) [reply pandas/_libs/lib.pyx]: Ah OK I see - thanks for clarifying

File: pandas/core/frame.py [hunk 5]
```diff
@@ -7820,34 +7820,30 @@ def _count_level(self, level, axis=0, numeric_only=False):
...
 
         if axis == 1:
-            # Undo our earlier transpose
-            return result.T
+            result = DataFrame(counts, index=agg_axis, columns=level_index)
         else:
-            return result
+            result = DataFrame(counts, index=level_index, columns=agg_axis)
+
+        return result
```
Comment (tv3141) [inline pandas/core/frame.py]: I amended the masking logic in this function to mask all rows/columns that `lib.count_level_2d` requires to be masked  to avoid out-of-bounds access.

**Comment by tv3141** [comment]
ping @jreback

**Review by jreback — APPROVED** [review]
lgtm. @jbrockmendel if any comments.

**Comment by WillAyd** [comment]
@tv3141 can you fix merge conflict in whatsnew? Otherwise lgtm as well

File: pandas/core/frame.py [hunk 6]
```diff
@@ -7890,34 +7890,30 @@ def _count_level(self, level, axis=0, numeric_only=False):
...
-        if frame._is_mixed_type:
-            # Since we have mixed types, calling notna(frame.values) might
-            # upcast everything to object
-            mask = notna(frame).values
-        else:
-            # But use the speedup when we have homogeneous dtypes
-            mask = notna(frame.values)
+        # Mask NaNs: Mask rows or columns where the index level is NaN, and all
+        # values in the DataFrame that are NaN
+        values_mask = notna(frame.values)
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: is the existing version with `notna(frame).values` wrong?  avoiding .values with is_mixed_type is worthwhile

Comment (tv3141) [reply pandas/core/frame.py]: @jbrockmendel I understood @jreback 's earlier [comment](https://github.com/pandas-dev/pandas/pull/32842#discussion_r396028090) such that `notna(frame.values)` and `notna(frame).values` are the same.

Comment (jbrockmendel) [reply pandas/core/frame.py]: those two are in fact the same, but performance should be better for `notna(frame).values`

Comment (tv3141) [reply pandas/core/frame.py]: Running some tests shows that `notna(df).values` is indeed much faster on larger mixed type `DataFrames`. I will restore the original code. 
I guess this optimisation would be better in `notna`.
```python
In [1]: import pandas as pd                                                                                                                                                                   
In [2]: from pandas.core.dtypes.missing import notna                                                                                                                               
In [3]: df = pd.DataFrame({'a': range(10), 'b': ['foo'] * 10})                                                                                                                                     
In [4]: %timeit notna(df.values)                                                                                                                                                              
113 µs ± 9.55 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
In [5]: %timeit notna(df).values                                                                                                                                                              
546 µs ± 5.22 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
In [6]: df = pd.DataFrame({'a': range(1000000), 'b': ['foo'] * 1000000})                                                                                                                      
In [7]: %timeit notna(df.values)                                                                                                                                                              
163 ms ± 2.35 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
In [8]: %timeit notna(df).values                                                                                                                                                              
40.7 ms ± 291 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [9]: %timeit notna(df.values)                                                                                                                                                              
158 ms ± 1.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
In [10]: %timeit notna(df).values                                                                                                                                                             
39.7 ms ± 1.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by WillAyd** [comment]
@jbrockmendel any comments outstanding on this?

**Comment by jreback** [comment]
this looks good. i think ok to merge.

**Comment by jbrockmendel** [comment]
no objections here

**Comment by WillAyd** [comment]
Great thanks @tv3141
